### DISCUSSION TITLE
Snapdragon Spaces OpenXR Custom Loader Setup

### DISCUSSION THREAD

**Discussion by maluoi** [StereoKit/StereoKit#790] [General]
Currently, Snapdragon Spaces powered OpenXR headsets don't use the standardized Khronos OpenXR loader. To use StereoKit on these devices, a little bit of additional setup is required. Eventually, Snapdragon Spaces may move to the official loader and these steps will no longer be necessary, and if so, this document will be updated to reflect that.

# Acquiring the Loader

To get the Snapdragon custom OpenXR loader, you need to grab the Unity SDK for Snapdragon Spaces (account required).
https://spaces.qualcomm.com/download-sdk/?lang=en

From the SDK, unzip the `Unity Package/SnapdragonSpaces_Unity_Samples_0_19_1.apk` APK file.

From the unzipped APK, you'll find `libopenxr_loader.so` in the `lib/arm64-v8a` folder. This is the file you need.

# Using the Loader

There's a few options for setting up StereoKit to use the loader:

## 1. Insert the files into the NuGet package

If you don't switch StereoKit version often, inserting the loader files into the NuGet package is probably one of the easier options!

The NuGet package for StereoKit is located approximately here, your version number may vary.
`C:\Users\[user]\.nuget\packages\stereokit\0.3.8`

You'll want to add the loader files you extracted like this (case sensitive):
```
-+ NuGet Folder
 |-+ android_openxr_loaders
   |-+ snapdragon
     |-+ arm64
       |-- libopenxr_loader.so
```

In your project, you can then open up `[Project Dir]/Projects/Android/[Project Name]_Android.csproj` and:
```xml
<!-- Change this: -->
<SKOpenXRLoader>Standard</SKOpenXRLoader>

<!-- To this -->
<SKOpenXRLoader>Snapdragon</SKOpenXRLoader>
```

## 2. OR, make a shared folder

This is useful if you switch StereoKit versions frequently.

Create a folder somewhere on your hard drive to store the loader files. I do this in `C:/Tools`, but this is merely my own preference:
```
-+ C
 |-+ Tools
   |-+ openxr_loaders
     |-+ Snapdragon
       |-+ arm64
         |-- libopenxr_loader.so
```

Then you can do this to your project file `[Project Dir]/Projects/Android/[Project Name]_Android.csproj` :
```xml
<!-- Change this: -->
<SKOpenXRLoader>Standard</SKOpenXRLoader>

<!-- To this -->
<SKOpenXRLoaderFolder>C:/Tools/openxr_loaders</SKOpenXRLoaderFolder>
<SKOpenXRLoader>Snapdragon</SKOpenXRLoader>
```

**Comment by cefoot** [comment]
With current Version 19.1 of the Spaces SDK the `libc++_shared.so` is no longer.
As mentioned in the Changelog:

- The C++ library inside the Android OpenXR loader archive has been changed with a static one to avoid conflicts with potential other plugin archives using the libc++_shared.so file. ⚠ Warning: This might break functionality with the Snapdragon Spaces Services prior to version 0.19.1.

The `libopenxr_loader.so` is still present, it is the only file necessary. The above-mentioned code change in the `MainActivity.cs` is also no longer needed.

Oh and I never made it work with the **_shared folder_** Option 2. But modifying the nuget package folder worked.

**Reply by maluoi** [reply]
Thanks for pointing this one out! 0.19.1 does indeed remove libc++_shared.so! I've confirmed it works well, and updated the post to reflect this :)

I'll take a peek at the shared folder bit, see if something's broken.

**Comment by Leuconoe** [comment]
This article has been very helpful to me.

**Comment by GLM-FM** [comment]
To add something I noticed by accident, StereoKit (whether by design or accident) also seems to look for an `openxr_loader` directory next to the Android project file without explicitly specifying SKOpenXRLoaderFolder.

Only thing to take care of is including this line in an ItemGroup so it isn't picked up by the regular Android native library handling of the .NET SDK:

`<AndroidNativeLibrary Remove="openxr_loader\**"/>`
